### PR TITLE
Correctly persist and send length & encoded_length values (3rd attempt)

### PR DESCRIPTION
**Pull request by tomblench**
What persist to the local DB the length and encoded_length metadata properties when they are sent with attachments.
Why currently this data is not persisted or used to check that the actual length equals the indicated length.
How update various Attachment classes and related methods to take length and encoded_length arguments and assert that actual attachment length equals signalled attachment length
Testing mostly covered by existing tests

reviewer @rhyshort 
reviewer @mikerhodes

### CODE REVIEW

**Comment by rhyshort** [comment]
+1, once the commit message reads something sensible

File: sync-core/src/main/java/com/cloudant/sync/datastore/PreparedAttachment.java [hunk 1]
```diff
@@ -81,16 +93,20 @@ public PreparedAttachment(Attachment attachment,
...
             IOUtils.closeQuietly(tempFileOutStream);
         }
+        
         //Set attachment length from bytes read in input stream
-        this.length = totalRead;
+        if (this.attachment.encoding == Attachment.Encoding.Plain) {
+            this.length = totalRead;
+            // 0 signals "no encoded length" - this is consistent with couch which does not send
+            // encoded_length if the encoding is "plain"
+            this.encodedLength = 0;
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/datastore/PreparedAttachment.java]: It'd be nice to use a constant for this, e.g., `NO_ENCODED_LENGTH`?

File: sync-core/src/main/java/com/cloudant/sync/replication/BasicPullStrategy.java [hunk 2]
```diff
@@ -291,7 +291,17 @@ private int processOneChangesBatch(ChangesResultWrapper changeFeeds)
                                 atts.put(new String[]{documentRevs.getId(), documentRevs.getRev()}, preparedAtts);
 
                                 for (String attachmentName : attachments.keySet()) {
-                                    int revpos = (Integer) ((Map<String, Object>) attachments.get(attachmentName)).get("revpos");
+                                    Map attachmentMetadata = (Map)attachments.get(attachmentName);
+                                    int revpos = (Integer) attachmentMetadata.get("revpos");
+                                    String contentType = (String) attachmentMetadata.get("content_type");
+                                    String encoding = (String) attachmentMetadata.get("encoding");
+                                    long length = (Integer) attachmentMetadata.get("length");
+                                    long encodedLength = 0; // encodedLength can default to 0 if it's not encoded
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/BasicPullStrategy.java]: Constant here?

**Comment by mikerhodes** [comment]
Looking good, add the constant if it makes sense. Otherwise +1.

File: sync-core/src/main/java/com/cloudant/sync/datastore/PreparedAttachment.java (hunk 1 above)
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/PreparedAttachment.java]: Agreed. The logic is hard to follow but I think we should be consistent with the naming and meaning of these values in CouchDB. So the constant will at least help with that.
